### PR TITLE
Bug 1318411 - Proxy should not strip querystring, also avoid hardcodi…

### PR DESCRIPTION
**Pull request by jonasfj**
…ng hostnames

This makes it so that:
```
http://taskcluster/auth/...    -> https://auth.taskcluster.net/...
http://taskcluster/myhost.example.com/... -> https://myhost.example.com/...
```

Generically:
```
http://taskcluster/<host>/<path>[?<query>]

if <host> contains a dot '.':
  https://<host>/<path>[?<query>]

if <host> doesn't contain a dot '.':
  https://<host>.taskcluster.net/<path>[?<query>]
```

The hawk HMAC signature contains the hostname, so a signed request to `myevil-service.ngrok.io` can't be used to obtain a request signature that can be exploited to create a task on the queue...
It does reveal the certificate from the temporary credentials, but those are useless with the accessToken, and technically not secret.

### CODE REVIEW

**Comment by petemoore** [comment]
@jonasfj In light of https://bugzilla.mozilla.org/show_bug.cgi?id=1310643#c12 (also comment 13) we should probably either revert the change to add a root handler, or just fix NSS not to have an extra trailing slash in the base url. Just wanted to raise this now before reviewing, as currently rolling out would break NSS. Will review properly when I'm back at work next week (although on PTO on Monday).

**Comment by petemoore** [comment]
@ttaubert @jonasfj I've made a fix for the NSS part of this [here](https://bug1310643.bmoattachments.org/attachment.cgi?id=8812447) so this unblocks this PR.

Be sure to follow [the release guidelines](https://github.com/taskcluster/taskcluster-proxy#making-a-release) when releasing this change.

**Comment by petemoore** [comment]
@jonasfj Note, it might not be a bad idea to update [taskcluster-client](https://github.com/taskcluster/taskcluster-client) to strip any trailing slash from `baseUrl` to avoid double slashes.

File: taskcluster/services.go [hunk 1]
```diff
@@ -3,54 +3,75 @@ package taskcluster
...
-		"scheduler":       "https://scheduler.taskcluster.net/",
-		"secrets":         "https://secrets.taskcluster.net/",
-	}
-
-	return Services{Endpoints: endpoints}
+	return Services{Domain: "taskcluster.net"}
 }
 
-// Convert a url for the proxy server into a url for the proper taskcluster
+var hostnamePattern = regexp.MustCompile(`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`)
```
Comment (petemoore) [inline taskcluster/services.go]: `^([a-zA-Z0-9](|[a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*[a-zA-Z0-9](|[a-zA-Z0-9\-]*[a-zA-Z0-9])$` is a bit shorter, but yours is also fine. :-)

File: taskcluster/services_test.go [hunk 2]
```diff
@@ -27,6 +27,22 @@ var urlConversions = []struct {
...
+	},
+	{
+		"https://xfoo.com/queue/x/y/z?key=value",
+		"https://queue.taskcluster.net/x/y/z?key=value",
+	},
+	{
+		"https://xfoo.com/myqueue.somewhere.com/v1/task/tsdtwe34tgs%2ff5yh?k=v%20m",
+		"https://myqueue.somewhere.com/v1/task/tsdtwe34tgs%2ff5yh?k=v%20m",
+	},
 }
```
Comment (petemoore) [inline taskcluster/services_test.go]: Maybe extra useful tests might be query strings including a `&`, maybe a url which contains other unicode characters (can a path contain non-ascii characters?). I'm also guessing it isn't possible to construct a url that is invalid UTF8, so maybe we don't need to test that too.

File: taskcluster/services.go (hunk 1 above)
Comment (jonasfj) [inline taskcluster/services.go]: I just stole it from stackoverflow...  :)

File: taskcluster/services_test.go (hunk 2 above)
Comment (jonasfj) [inline taskcluster/services_test.go]: invalid hostname are tested in the integration tests... But `&` would be good.
